notifications_utils/field.py

```diff
@@ -91,6 +91,17 @@ def __init__(
             'escape': escape_html,
             'passthrough': str,
         }[html]
+
+        # html special characters in placeholders are escaped in certain cases(depending on the
+        # sanitizer configuration) which results in a divergence from the un-sanitized
+        # placeholder stored in self.values.
+        # sanitized_values is being introduced to address this issue. It will now be used by
+        # get_replacement method to get the value to replace the placeholder.
+        if self.values:
+            self.sanitized_values = InsensitiveDict({
+                self.sanitizer(k): v for k, v in self.values.items()
+            })
```

Comment:
You probably don’t need the `if` here. If `values` is falsey then it gets coerced to `{}` here:
https://github.com/alphagov/notifications-utils/blob/79cc8da3e1f871d269b6c6f85d11ac20c3858e98/notifications_utils/field.py#L113

I suspect that `self.sanitized_values = InsensitiveDict({})` would be optimised to take negligible time.

That would look like:
```suggestion
        self.sanitized_values = InsensitiveDict({
            self.sanitizer(k): v for k, v in self.values.items()
        })
```

***

You could also rewrite it as a cached property, meaning its only computed when its needed, not on every `__init__`:
```python
class Field:

    […]

    @cached_property
    def sanitized_values(self):
        return InsensitiveDict({
            self.sanitizer(k): v for k, v in self.values.items()
        })
```

***

And if you wanted to go one step further you could subclass `InsensitiveDict` so that you don’t have to sanitise every key:

```python
class SanitisedKeyInsensitiveDict(InsensitiveDict):

    def __init__(self, row_dict, sanitiser):
        self._sanitiser = sanitiser

    @staticmethod
    def make_key(original_key):
        return self.sanitiser(super().make_key(original_key))

class Field:

    […]

    @cached_property
    def sanitized_values(self):
        return SanitisedKeyDict(self.values, sanitiser=self.sanitiser)
```

***

I can show you how to performance test this code if you like, and work out if any of those optimisations are worth doing.